pip/pip-369.md

```diff
@@ -0,0 +1,121 @@
+
+And now in 3.3, `enableNamespaceIsolationUpdateOnTime` broker config was deprecated as it no longer serves any purpose. https://github.com/apache/pulsar/pull/22449
+
+# Motivation
+
+In Apache Pulsar 3.x, changing a namespace isolation policy results in unloading all namespace bundles that match the namespace's regular expression provided in the isolation policy.
+This can be problematic for cases where the regex matches a large subset of namespaces, such as `tenant-x/.*`. 
+One of such case is mentioned on this issue [#23092](https://github.com/apache/pulsar/issues/23092) where policy change resulted in 100+ namespace bundles to get unloaded.
+And broker exhausted all the available connections due to too many unload calls happening at once resulting in 5xx response.
+Other issues that happens with this approach are huge latency spikes as topics are unavailable until bundles are loaded back, increasing the pending produce calls.
+The only benefit this approach serves is ensuring that all the namespaces matching the policy regex will come to correct broker group. 
+But when namespace bundles are already on the correct broker group (according to the policy), unloading those namespaces doesn't serve any purpose.
+
+This PIP aims to address the need to either prevent unnecessary unloading or provide a more granular approach to determine what should be unloaded.
+
+Some of the cases covered by this PIP are discussed in [#23094](https://github.com/apache/pulsar/issues/23094) by @grssam.
+> - unload nothing as part of the set policy call
+> - unload every matching namespace as part of the policy set call
+> - unload only the changed namespaces (newly added + removed)
+
+# Goals
+
+## In Scope
+This PIP proposes a flag-based approach to control what should be unloaded when an isolation policy is applied.
+The possible values for this flag are:
+- **all**: Unload all the namespaces that matches either old or new policy change.
+- **changed**: Only unload namespaces that are either added or removed due to the policy change.
+- **none**: Do not unload anything. Unloading can occur naturally due to load balancing or can be done manually using the unload admin call.
+
+This flag will be a part of isolation policy data with defaults. Objective is to keep the default behavior unchanged on applying the new policy.
+
+## Out of Scope
+
+Applying concurrency reducer to limit how many async calls will happen in parallel is out of the scope for this PIP. 
+This should be addressed in a separate PIP, as solving the issue of infinite asynchronous calls probably requires changes to broker configurations and is a problem present in multiple areas.
+
+# Detailed Design
+
+## Design & Implementation Details
+
+A new flag will be introduced in `NamespaceIsolationData`.
+
+```java
+enum unload {
+    all,        // unloads everything, OLD ⋃ NEW
```

Comment:
Are there any benefits to using this flag? If not, how about removing this option?

Reply:
@RobertIndie - Are you referring about just the "all" option of the "unload" flag?

Reply:
No. I'm suggesting removing the `all` option like this:
```java
enum unload {
    changed,    // unload namespaces delta, (new ⋃ old) - (new ∩ old) 
    none,       // skip unloading anything, ϕ
};
```

Reply:
that would be backward incompatible as since pulsar 2.7, it is doing what the `all` is meant to do.
We can discuss removing it in 4.x where we change the default to `changed`

Reply:
Yeah.  What's the difference between `all` and `changed`? It seems that `all` offers no clear benefits, right? Or may I miss something? IMO, to users, this won't change the expected behavior, so we might not need to support this option.

Besides, there is a discussion for the 4.0.0 release: https://lists.apache.org/thread/mbrxy6jnz1yj0pyhmlr39wy0w1dhqv7s. The next feature release may be 4.0.0 instead of 3.4.0.

Reply:
`all` - all may fix all namespaces without making any assumptions. using the `all` option means that it ensures both the older and newer namespaces are in the right locattion.
`changed` takes an assumption that the unchanged namespaces are already in the correct location so it only unloads the namespaces which are either newly added or removed now.

Based on how things are, users can take a call to have force place every namespace or not.

Regarding releasing this - We want this to be present in 3.x as well since this majorly is a degradation in 2.11/3.x from 2.7/2.9 where there _was_ an option to disable this "unload all namespace" behavior. I believe all production systems will benefit from this present in the current LTS version, rather than making it wait for october time frame for 4.0 release.

The implementation PR is ready as well.

Reply:
Thank you for clarifying. I had misunderstood the terms "new" and "old" before. Now I get it.

>  I believe all production systems will benefit from this present in the current LTS version

So you mean this PIP will be cherry-picked to the LTS version of 3.0.x?  I think we should discuss this separately or state it in the proposal. As we only allow bug fixes and important improvements cherry-picked into the released LTS version. 

And finally, I'm +1 for this PIP.

Reply:
> Thank you for clarifying. I had misunderstood the terms "new" and "old" before. Now I get it.

There is a catch for 3.x vs 4.0 in the meaning of "all" mentioned in the notes, again for compatibility reasons.

> So you mean this PIP will be cherry-picked to the LTS version of 3.0.x? I think we should discuss this separately or state it in the proposal. As we only allow bug fixes and important improvements cherry-picked into the released LTS version.

That's the idea. We will add it explicitly in this PIP proposal itself. 

> And finally, I'm +1 for this PIP.
Thank you for reviewing!